### DISCUSSION TITLE
How could I use a custom Backend like Supabase?

### DISCUSSION THREAD

**Discussion by Cherryrubim** [firecmsco/firecms#453] [Q&A]
I've been reading the documentation for custom backend, but I'm still unclear on how I could extend or replace the logic for the `DataSource`, `StorageSource`, and `AuthController` interfaces.

I'll be grateful for any information or assistance provided. 🙏

**Comment by fgatti675** [comment]
Hi @Cherryrubim 

It's very cool that you want to implement a backend like Supabase!
It is something we've had in our minds for a while.

For that reason, could you join our Discord channel so we can discuss the details?
https://discord.com/invite/fxy7xsQm3m

In any case, as you said, you need to implement all the non-optional fields of those interfaces.

Internally we have implemented those components using React Hooks:
- Auth : https://github.com/Camberi/firecms/blob/main/lib/src/firebase_app/hooks/useFirebaseAuthController.ts
This one is tied to our [login view](https://github.com/Camberi/firecms/blob/main/lib/src/firebase_app/components/FirebaseLoginView.tsx)
Essentially in charge of returning a logged in user.

- Datasource: https://github.com/Camberi/firecms/blob/main/lib/src/firebase_app/hooks/useFirestoreDataSource.ts
For saving and retrieving data.

- Storage: https://github.com/Camberi/firecms/blob/main/lib/src/firebase_app/hooks/useFirebaseStorageSource.ts
For uploading files.

Then you can use them in a custom app like in this example:
https://firecms.co/docs/custom_cms_app

Let me know if you need any help!!
